quantum/qauntum_logic_gates.py

```diff
@@ -0,0 +1,381 @@
+    of the input complex number.
+    In cases which require 2 inputs the input is named with a suffix of 1 and 2
+    (Eg. input_realvalue_1)
+
+    Usage :
+    >>> controlled_phase_multiplication(3,2,5,1,10)
+    array([3.00000000e+00+0.j, 2.00000000e+00+0.j, 1.10132329e+05+0.j,
+           2.20264658e+04+0.j])
+    """
+    iota = cmath.exp(alpha)
+    controlled_phase_multiplication_matrix = np.array(
+        [[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, iota, 0], [0, 0, 0, iota]]
+    )
+    complex_input = np.array(
+        [
+            input_realvalue_1,
+            input_imaginaryvalue_1,
+            input_realvalue_2,
+            input_imaginaryvalue_2,
+        ]
+    )
+    result = np.dot(controlled_phase_multiplication_matrix, complex_input)
+    return result
+
+
+def swap_gate(
+    input_realvalue_1: float,
+    input_imaginaryvalue_1: float,
+    input_realvalue_2: float,
+    input_imaginaryvalue_2: float,
+) -> list[float]:
+    """
+    Glossary :
+    input_realvalue : the magnitude of the real part of the input complex number.
+    input_imaginaryvalue : the magnitude of the imaginary part
+    of the input complex number.
+    In cases which require 2 inputs the input is named with a suffix of 1 and 2
+    (Eg. input_realvalue_1)
+
+    Usage :
+    >>> swap_gate(5,1,3,7)
+    array([5, 3, 1, 7])
+    """
+    swap_gate_matrix = np.array(
+        [[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 0, 1]]
+    )
+    complex_input = np.array(
+        [
+            input_realvalue_1,
+            input_imaginaryvalue_1,
+            input_realvalue_2,
+            input_imaginaryvalue_2,
+        ]
+    )
+    result = np.dot(swap_gate_matrix, complex_input)
+    return result
+
+
+def spin_of_input(
+    input_realvalue: float,
+    input_imaginaryvalue: float,
+    alpha_value: float,
+    nx_value: float,
+    ny_value: float,
+    nz_value: float,
+) -> list[float]:
+    """
+    Glossary :
+    alpha : angle of rotation as represented by the block sphere.
+    iota : The exponential complex of alpha value.
+    nx_value : value of vector in X axis as represented by Hilbert space.
+    nx_value : value of vector in Y axis as represented by Hilbert space.
+    nx_value : value of vector in Z axis as represented by Hilbert space.
+
+    * The nx,ny and nz values can also be considered as values of vectors along
+    the respective axes on the bloch sphere.
+
+    Usage :
+    >>> spin_of_input(6,3,45,1,8,3)
+    array([-16.93201614+10.23066476j, -50.61991392 -1.46152354j])
+    """
+    i = complex(0, 1)
+    spin_matrix = [
+        [
+            (math.cos(alpha_value / 2.0))
+            - (i * math.sin(alpha_value / 2.0) * nz_value),
+            (-1 * i * math.sin(alpha_value / 2.0) * (nx_value + i * ny_value)),
+        ],
+        [
+            -1 * i * (math.sin(alpha_value / 2.0) * nx_value - i * ny_value),
+            math.cos(alpha_value / 2.0) + (i * math.sin(alpha_value / 2.0) * nz_value),
+        ],
+    ]
```

Comment:
```suggestion
    spin_matrix = [
        [
            math.cos(alpha_value / 2) - 1j * math.sin(alpha_value / 2) * nz_value),
            -1j * math.sin(alpha_value / 2) * (nx_value - 1j * ny_value),
        ],
        [
            -1j * math.sin(alpha_value / 2) * (nx_value + 1j * ny_value),
            math.cos(alpha_value / 2) + 1j * math.sin(alpha_value / 2) * nz_value,
        ],
    ]
```
1) Use complex literals with `j`
2) I believe the spin matrix here is incorrect. Scherer has this as the spin matrix:

$$
\begin{pmatrix}
 \cos \left( \frac{\alpha}{2} \right) - i \sin \left( \frac{\alpha}{2} \right) n_z & -i \sin \left( \frac{\alpha}{2} \right) (n_x - i n_y)\\
 -i \sin \left( \frac{\alpha}{2} \right) (n_x + i n_y) & \cos \left( \frac{\alpha}{2} \right) + i \sin \left( \frac{\alpha}{2} \right) n_z
\end{pmatrix}
$$

whereas you have this as the spin matrix:

$$
\begin{pmatrix}
 \cos \left( \frac{\alpha}{2} \right) - i \sin \left( \frac{\alpha}{2} \right) n_z & -i \sin \left( \frac{\alpha}{2} \right) (n_x + i n_y)\\
 -i \sin \left( \frac{\alpha}{2} \right) (n_x - i n_y) & \cos \left( \frac{\alpha}{2} \right) + i \sin \left( \frac{\alpha}{2} \right) n_z
\end{pmatrix}
$$

In your matrix, the top-right and bottom-left entries are switched.